<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bioinformatics Alignment Pandarallel Nucleotides and Amino Acids Benchmark in Single and Multi-node Environments &mdash; Dragon  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Dragon
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../start/start.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../uguide/uguide.html">Users Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pguide/pguide.html">Programming Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="cbook.html">Solution Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../benchmarks/benchmarks.html">Benchmarks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq/faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ref/ref.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../infrastructure/infrastructure.html">Infrastructure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../services/services.html">Services</a></li>
<li class="toctree-l1"><a class="reference internal" href="../components/components.html">Low-level Components</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Dragon</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Bioinformatics Alignment Pandarallel Nucleotides and Amino Acids Benchmark in Single and Multi-node Environments</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/cbook/bioinfo_alignment_pandarallel_demo.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bioinformatics-alignment-pandarallel-nucleotides-and-amino-acids-benchmark-in-single-and-multi-node-environments">
<h1>Bioinformatics Alignment Pandarallel Nucleotides and Amino Acids Benchmark in Single and Multi-node Environments<a class="headerlink" href="#bioinformatics-alignment-pandarallel-nucleotides-and-amino-acids-benchmark-in-single-and-multi-node-environments" title="Permalink to this heading"></a></h1>
<p>This Jupyter benchmark performs both DNA and protein alignments in parallel using the
pandarallel <code class="code docutils literal notranslate"><span class="pre">parallel_apply</span></code> call. It can be run with <code class="code docutils literal notranslate"><span class="pre">dragon</span></code> to
compare performance on your machine. The DNA/nucleotide workload is run in a single-node environment.
The protein/amino acid workload is run in a multi-node environment.</p>
<p>The use case utilizes pairwise alignments from pyalign, a jaccard distance calculation for the E value, and a hamming distance calculation for the coverage percentage.
The timings are provided with Dragon and base multiprocessing for <code class="code docutils literal notranslate"><span class="pre">parallel_apply</span></code>, the multiprocessing verison of pandas <code class="code docutils literal notranslate"><span class="pre">apply</span></code>.
The application utilizes nucleotide and amino acid workloads for feature selection.
The time to run the workloads is calculated and displayed in the pandas. K-means clustering is used to group sequences by alignment and percentage coverage of alignments.</p>
<p>The code demonstrates the following key concepts working with Dragon:</p>
<ul class="simple">
<li><p>How to write programs that can run with Dragon and base multiprocessing</p></li>
<li><p>How to use pandarallel and pandas with Dragon for feature selection</p></li>
<li><p>How pandarallel handles different dtypes</p></li>
<li><p>How to utilize pandarallel in a multi-node environment</p></li>
<li><p>How to utilize k-means clustering on features such as alignment, E value, and percentage coverage</p></li>
</ul>
<div class="literal-block-wrapper docutils container" id="id1">
<div class="code-block-caption"><span class="caption-text"><strong>bioinformatics_alignment_pandarallel_demo.ipynb: A bioinformatics benchmark for aligning nucleotide sequences and amino acid sequences</strong></span><a class="headerlink" href="#id1" title="Permalink to this code"></a></div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="linenos"> 1</span><span class="kn">import</span> <span class="nn">dragon</span>
<span class="linenos"> 2</span><span class="kn">import</span> <span class="nn">multiprocessing</span>
<span class="linenos"> 3</span>
<span class="linenos"> 4</span><span class="kn">import</span> <span class="nn">cloudpickle</span>
<span class="linenos"> 5</span>
<span class="linenos"> 6</span><span class="kn">import</span> <span class="nn">os</span>
<span class="linenos"> 7</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;OPENBLAS_NUM_THREADS&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;1&#39;</span>
<span class="linenos"> 8</span>
<span class="linenos"> 9</span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="linenos">10</span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="linenos">11</span>
<span class="linenos">12</span><span class="kn">import</span> <span class="nn">Bio</span>
<span class="linenos">13</span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span><span class="p">,</span> <span class="n">Entrez</span>
<span class="linenos">14</span><span class="kn">import</span> <span class="nn">pyalign</span>
<span class="linenos">15</span><span class="kn">import</span> <span class="nn">time</span>
<span class="linenos">16</span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="linenos">17</span><span class="kn">from</span> <span class="nn">sklearn.cluster</span> <span class="kn">import</span> <span class="n">KMeans</span>
<span class="linenos">18</span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="linenos">19</span><span class="kn">import</span> <span class="nn">pandarallel</span><span class="p">;</span> <span class="n">pandarallel</span><span class="o">.</span><span class="n">__version__</span>
<span class="linenos">20</span>
<span class="linenos">21</span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">set_start_method</span><span class="p">(</span><span class="s2">&quot;dragon&quot;</span><span class="p">)</span>
<span class="linenos">22</span><span class="n">pandarallel</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">dill</span> <span class="o">=</span> <span class="n">cloudpickle</span>
<span class="linenos">23</span><span class="n">pandarallel</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">CONTEXT</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">get_context</span><span class="p">(</span><span class="s2">&quot;dragon&quot;</span><span class="p">)</span>
<span class="linenos">24</span><span class="n">pandarallel</span><span class="o">.</span><span class="n">pandarallel</span><span class="o">.</span><span class="n">initialize</span><span class="p">(</span><span class="n">progress_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos">25</span>
<span class="linenos">26</span><span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">monotonic</span><span class="p">()</span>
<span class="linenos">27</span><span class="n">nucl_df</span><span class="p">[</span><span class="s1">&#39;PyAlign Alignment Score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">nucl_df</span><span class="p">[</span><span class="s1">&#39;Sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">parallel_apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">seq2</span><span class="p">:</span> <span class="n">alignment_algorithm</span><span class="p">(</span><span class="n">endo_nucl_seq</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">gap</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
<span class="linenos">28</span><span class="n">stop</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">monotonic</span><span class="p">()</span>
<span class="linenos">29</span><span class="n">functions</span><span class="p">,</span> <span class="n">bar_num</span><span class="p">,</span> <span class="n">tot_time</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PyAlign Alignment Score&#39;</span><span class="p">],[</span><span class="mi">128</span><span class="p">],[</span><span class="n">stop</span><span class="o">-</span><span class="n">start</span><span class="p">]</span>
</pre></div>
</div>
</div>
<p>For the single-node run, both base multiprocessing and Dragon are compared. The runs utilized a single node with 2 AMD EPYC 7742 64-Core Processors with 128 cores.
Dragon employs a number of optimizations on base multiprocessing; the Dragon start method outperforms the use of the base multiprocessing spawn start method on the same hardware.</p>
<p>The timing for the base multiprocessing runtime is:</p>
<table class="colwidths-given docutils align-default" id="id2">
<caption><span class="caption-text">Base Multiprocessing Timings for Nucleotide Alignments with Different Number of Bars</span><a class="headerlink" href="#id2" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Pandarallel Function</p></th>
<th class="head"><p>Number of Bars</p></th>
<th class="head"><p>Time</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>PyAlign Alignment Score</p></td>
<td><p>128</p></td>
<td><p>61.000052</p></td>
</tr>
<tr class="row-odd"><td><p>E Score</p></td>
<td><p>10</p></td>
<td><p>22.919291</p></td>
</tr>
<tr class="row-even"><td><p>Percentage Coverage</p></td>
<td><p>1</p></td>
<td><p>18.000021</p></td>
</tr>
<tr class="row-odd"><td><p>Total Time</p></td>
<td></td>
<td><p>101.919364</p></td>
</tr>
</tbody>
</table>
<p>The timing for the single-node Dragon runtime is:</p>
<table class="colwidths-given docutils align-default" id="id3">
<caption><span class="caption-text">Dragon Timings for Nucleotide Alignments with Different Number of Bars</span><a class="headerlink" href="#id3" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Pandarallel Function</p></th>
<th class="head"><p>Number of Bars</p></th>
<th class="head"><p>Time</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>PyAlign Alignment Score</p></td>
<td><p>128</p></td>
<td><p>11.601343</p></td>
</tr>
<tr class="row-odd"><td><p>E Score</p></td>
<td><p>10</p></td>
<td><p>7.882140</p></td>
</tr>
<tr class="row-even"><td><p>Percentage Coverage</p></td>
<td><p>1</p></td>
<td><p>7.930996</p></td>
</tr>
<tr class="row-odd"><td><p>Total Time</p></td>
<td></td>
<td><p>27.174203</p></td>
</tr>
</tbody>
</table>
<p>For multi-node Dragon run, the run was on 2 Apollo nodes. Each Apollo node has 1x AMD Rome CPU with 4x AMD MI100 GPUs and 128 cores.
The multi-node use case scales with the total number of CPUs reported by the allocation. As there are more nodes, workers, and CPUs available for multi-node, Dragon extends
multiprocessing’s stock capabilities and demonstrates additional improvement to measured execution time.
Base multiprocessing does not support multi-node workloads.</p>
<p>The timing for the multi-node Dragon runtime is:</p>
<table class="colwidths-given docutils align-default" id="id4">
<caption><span class="caption-text">Multi-node Timings for Amino Acids Alignments with Same Number of Bars</span><a class="headerlink" href="#id4" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Pandarallel Function</p></th>
<th class="head"><p>Number of Bars</p></th>
<th class="head"><p>Time</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>PyAlign Alignment Score</p></td>
<td><p>10</p></td>
<td><p>7.031509</p></td>
</tr>
<tr class="row-odd"><td><p>E Score</p></td>
<td><p>10</p></td>
<td><p>6.835784</p></td>
</tr>
<tr class="row-even"><td><p>Percentage Coverage</p></td>
<td><p>10</p></td>
<td><p>7.396781</p></td>
</tr>
<tr class="row-odd"><td><p>Total Time</p></td>
<td></td>
<td><p>21.264074</p></td>
</tr>
</tbody>
</table>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Hewlett Packard Enterprise.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>